internal/encoding/asn1/asn1.go

```diff
@@ -0,0 +1,219 @@
+		if err != nil {
+			return nil, err
+		}
+		if isPrimitive(identifier) {
+			// primitive value
+			primitiveNode := &primitiveValue{
+				identifier: identifier,
+				content:    content,
+			}
+			node.members = append(node.members, primitiveNode)
+		} else {
+			// constructed value
+			constructedNode := &constructedValue{
+				identifier: identifier,
+				rawContent: content,
+			}
+			node.members = append(node.members, constructedNode)
+
+			// add a new constructed node to the stack
+			valueStack = append(valueStack, constructedNode)
+		}
+	}
+	return rootConstructed, nil
+}
+
+// decodeMetadata decodes the metadata of a BER-encoded ASN.1 value.
+//
+// r is the input byte slice.
+// The first return value is the identifier octets.
+// The second return value is the content octets.
+// The third return value is the subsequent octets after the value.
+func decodeMetadata(r []byte) ([]byte, []byte, []byte, error) {
+	identifier, r, err := decodeIdentifier(r)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	contentLen, r, err := decodeLength(r)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	if contentLen > len(r) {
+		return nil, nil, nil, ErrEarlyEOF
+	}
+	return identifier, r[:contentLen], r[contentLen:], nil
+}
+
+// decodeIdentifier decodes decodeIdentifier octets.
+//
+// r is the input byte slice.
+// The first return value is the identifier octets.
+// The second return value is the subsequent value after the identifiers octets.
+func decodeIdentifier(r []byte) ([]byte, []byte, error) {
+	if len(r) < 1 {
+		return nil, nil, ErrEarlyEOF
+	}
+	offset := 0
+	b := r[offset]
+	offset++
+
+	// high-tag-number form
+	if b&0x1f == 0x1f {
+		for offset < len(r) && r[offset]&0x80 == 0x80 {
+			offset++
+		}
+		if offset >= len(r) {
+			return nil, nil, ErrEarlyEOF
+		}
+		offset++
+	}
+	return r[:offset], r[offset:], nil
+}
+
+// decodeLength decodes length octets.
+// Indefinite length is not supported
+//
+// r is the input byte slice.
+// The first return value is the length.
+// The second return value is the subsequent value after the length octets.
+func decodeLength(r []byte) (int, []byte, error) {
+	if len(r) < 1 {
+		return 0, nil, ErrEarlyEOF
+	}
+	offset := 0
+	b := r[offset]
+	offset++
+
+	if b < 0x80 {
+		// short form
+		return int(b), r[offset:], nil
+	} else if b == 0x80 {
+		// Indefinite-length method is not supported.
+		return 0, nil, ErrUnsupportedIndefinedLength
+	}
+
+	// long form
+	n := int(b & 0x7f)
+	if n > 4 {
+		// length must fit the memory space of the int type.
+		return 0, nil, ErrUnsupportedLength
+	}
+	if offset+n >= len(r) {
+		return 0, nil, ErrEarlyEOF
+	}
+	var length uint64
+	for i := 0; i < n; i++ {
+		length = (length << 8) | uint64(r[offset])
+		offset++
+	}
+	if length > uint64(math.MaxInt64) {
+		// double check in case that length is over 31 bits.
+		return 0, nil, ErrUnsupportedLength
+	}
+	return int(length), r[offset:], nil
```

Comment:
`length` is finally converted to `int`. Since `int` is either `32` or `64` bits in golang, we don't support `length` over `31` bits. Therefore, line 208 is not quite correct. You may consider `(length >> 31) > 0` or `math.MaxInt32`.

Reply:
Why 32/31 bits? A CMS might be over `2^16` (64 KiB) but won't be over `2^32`.

Reply:
Updated to use `math.MaxInt32`

The comment was carried from the v1.0.0-rc.1. I think that is to say the length should not be over 2^31 - 1 (2147483647).
```go
	if length < 0 {
		// double check in case that length is over 31 bits.
		return 0, ErrUnsupportedLength
	}
```